### PR TITLE
[FLINK-21386][datastream] Postpone FromElementsFunction serialization to respect later type customization

### PR DESCRIPTION
**Pull request by kezhuw**
## What is the purpose of the change
Respect type customization(eg. `SingleOutputStreamOperator.returns`) for `FromElementsFunction`.

## Brief change log
* Add test to assert that `AbstractUdfStreamOperator` redirects `OutputTypeConfigurable` to function
* Postpone `FromElementsFunction` serialization to respect later type customization

## Verifying this change

This change added tests and can be verified as follows:
* `StreamGraphGeneratorTesttestOutputTypeConfigurationWithUdfStreamOperator`
* `FromElementsFunctionTest`
* Existing tests that using `FromElementsFunction` directly or indirectly.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes**)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

## **WARNING: possible breaking change**
Previously, `FromElementsFunction` serializes elements eagerly. After this pr, it serializes elements lazily. Any code which depend on previous behavior may break. Depends on perspective, this might be treated as a breaking change. Following are possible affected apis:
* `StreamExecutionEnvironment.fromElements(OUT... data)`
* `StreamExecutionEnvironment.fromElements(Class<OUT> type, OUT... data)`
* `StreamExecutionEnvironment.fromCollection(Collection<OUT> data)`
* `StreamExecutionEnvironment.fromCollection(Collection<OUT> data, TypeInformation<OUT> typeInfo)`

### CODE REVIEW

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/graph/StreamGraphGeneratorTest.java [hunk 1]
```diff
@@ -209,6 +211,25 @@ public void testVirtualTransformations() throws Exception {
                         instanceof ShufflePartitioner);
     }
 
+    @Test
+    public void testOutputTypeConfigurationWithUdfStreamOperator() throws Exception {
```
Comment (dawidwys) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/graph/StreamGraphGeneratorTest.java:215]: How is this test related to the `fromElements` function?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java [hunk 2]
```diff
@@ -1096,12 +1096,7 @@ private StateBackend loadStateBackend(ReadableConfig configuration, ClassLoader
         // must not have null elements and mixed elements
         FromElementsFunction.checkCollection(data, typeInfo.getTypeClass());
 
-        SourceFunction<OUT> function;
-        try {
-            function = new FromElementsFunction<>(typeInfo.createSerializer(getConfig()), data);
-        } catch (IOException e) {
-            throw new RuntimeException(e.getMessage(), e);
-        }
+        SourceFunction<OUT> function = new FromElementsFunction<>(data);
```
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java:1099]: Could we add a single positive test for the `returns` method on `fromElements`?

**Review by dawidwys — CHANGES_REQUESTED** [review]
Thank you @kezhuw for the quick PR. I had some smaller comments.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/functions/FromElementsFunctionTest.java [hunk 3]
```diff
@@ -68,6 +88,135 @@ public void testStrings() {
...
+
+        assertNotEquals(existingSerializer, newSerializer);
+
+        List<String> result = runSource(source);
+
+        assertEquals(STRING_LIST_DATA, result);
+    }
+
+    @Test
+    public void testSetOutputTypeWithExistingBrokenSerializer() throws Exception {
```
[the remark this answers is not in the document]
Comment (kezhuw) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/api/functions/FromElementsFunctionTest.java:176]: `testSetOutputTypeWithExistingBrokenSerializer` will breaks if there is no `setOutputType` after while `testSetOutputTypeWithDifferentSerializer` will succeed(assumed that there is not assertion for different serializer). 

`testSetOutputTypeWithExistingBrokenSerializer` matches scenario @zentol posted in FLINK-21386 that eagerly determined is incorrect. I reuse `DeserializeTooMuchType` for its incorrectness(see `testDeSerializationError`).

There are do duplication between the two. How about removing `testSetOutputTypeWithDifferentSerializer` and adding  assertion for different serializer in `testSetOutputTypeWithExistingBrokenSerializer` ?

[the remark this answers is not in the document]
Comment (kezhuw) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/api/functions/FromElementsFunctionTest.java:151]: Merged to one `testSetOutputTypeWithExistingBrokenSerializer`.

**Comment by kezhuw** [comment]
Hi @dawidwys, I have pushed fixup commit for next reviewing cycle. Please confirm whether I have resolve your concerns. Thanks.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java (hunk 2 above)
Comment (kezhuw) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java:1099]: Add `StreamExecutionEnvironmentTest .testFromElementsPostConstructionType`.

**Comment by kezhuw** [comment]
Test was cancelled in [kafka_gelly](https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=16149&view=logs&j=c5f0071e-1851-543e-9a45-9ac140befc32).

**Review by dawidwys — APPROVED** [review]
Thanks for the update!. Looks good now! Merging...
